File: data_structures/3_python/b_linkedLists.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    # Check if the linked list is empty.
    def is_empty(self):
        return self.head is None

    # Get the first node in the linked list.
    def first(self):
        return self.head

    # Get the last node in the linked list.
    def last(self):
        return self.tail

    # Append a value to the end of the linked list.
    def append(self, value):
        new_node = Node(value)

        if self.tail:
            self.tail.next = new_node
        else:
            self.head = new_node

        self.tail = new_node

    # Prepend a value to the beginning of the linked list.
    def prepend(self, value):
        new_node = Node(value)

        if self.head:
            new_node.next = self.head
        else:
            self.tail = new_node

        self.head = new_node

    # Check if a value exists in the linked list.
    def contains(self, value):
        current = self.head

        while current:
            if current.value == value:
                return True
            current = current.next

        return False

    # Get the node at a specific index in the linked list.
    def node_at_index(self, index):
        current = self.head
        current_index = 0

        while current and current_index < index:
            current = current.next
            current_index += 1

        return current

    # Insert a value at a specific index in the linked list.
    def insert(self, value, index):
        if index < 0:
            return

        new_node = Node(value)

        if index == 0:
            self.prepend(value)
            return

        prev = self.node_at_index(index - 1)
        next_node = prev.next if prev else None

        if prev:
            prev.next = new_node

        if next_node:
            new_node.next = next_node

        if not next_node:
            self.tail = new_node

    # Remove the node at a specific index in the linked list.
    def remove(self, index):
        if index < 0:
            return

        if index == 0:
            if self.head:
                self.head = self.head.next
                if not self.head:
                    self.tail = None
            return

        prev = self.node_at_index(index - 1)
        if prev and prev.next:
            prev.next = prev.next.next

        if prev and not prev.next:
            self.tail = prev

    # Print the linked list.
    def print_list(self):
        current = self.head

        while current:
            print(f"{current.value} -> ", end="")
            current = current.next

        print("None")

    # Reverse the linked list.
    def reverse(self):
        prev = None
        current = self.head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        self.tail, self.head = self.head, prev

    # Return the number of nodes in the linked list.
    def count(self):
        count = 0
        current = self.head

        while current:
            count += 1
            current = current.next

        return count

    # Remove all nodes from the linked list.
    def remove_all(self):
        self.head = None
        self.tail = None

    # Return a list representation of the linked list.
    def to_list(self):
        result = []
        current = self.head

        while current:
            result.append(current.value)
            current = current.next

        return result

    # Remove all occurrences of a value from the linked list.
    def remove_all_occurrences(self, value):
        current = self.head
        prev = None

        while current:
            if current.value == value:
                if prev:
                    prev.next = current.next
                    current = current.next
                else:
                    self.head = self.head.next
                    current = self.head
            else:
                prev = current
                current = current.next

        if not self.head:
            self.tail = None


# Test the LinkedList class
if __name__ == "__main__":
    linked_list = LinkedList()
    linked_list.append(1)
    linked_list.append(2)
    linked_list.append(3)

    print("Linked List:")
    linked_list.print_list()

File: data_structures/1_cpp/b_linkedLists.cpp
#include <iostream>
#include <vector>

template <typename T>
class Node {
public:
    T value;
    Node* next;
    
    Node(T val) : value(val), next(nullptr) {}
};

template <typename T>
class LinkedList {
private:
    Node<T>* head;
    Node<T>* tail;

public:
    LinkedList() : head(nullptr), tail(nullptr) {}

    // Check if the linked list is empty.
    bool isEmpty() const {
        return head == nullptr;
    }

    // Get the first node in the linked list.
    Node<T>* first() const {
        return head;
    }

    // Get the last node in the linked list.
    Node<T>* last() const {
        return tail;
    }

    // Append a value to the end of the linked list.
    void append(T value) {
        Node<T>* newNode = new Node<T>(value);

        if (tail != nullptr) {
            tail->next = newNode;
        } else {
            head = newNode;
        }

        tail = newNode;
    }

    // Prepend a value to the beginning of the linked list.
    void prepend(T value) {
        Node<T>* newNode = new Node<T>(value);

        if (head != nullptr) {
            newNode->next = head;
        } else {
            tail = newNode;
        }

        head = newNode;
    }

    // Check if a value exists in the linked list.
    bool contains(T value) const {
        Node<T>* current = head;

        while (current != nullptr) {
            if (current->value == value) {
                return true;
            }
            current = current->next;
        }

        return false;
    }

    // Get the node at a specific index in the linked list.
    Node<T>* nodeAtIndex(int index) const {
        if (index < 0)
            return nullptr;

        Node<T>* current = head;
        int currentIndex = 0;

        while (current != nullptr && currentIndex < index) {
            current = current->next;
            currentIndex++;
        }

        return current;
    }

    // Insert a value at a specific index in the linked list.
    void insert(T value, int index) {
        if (index < 0)
            return;

        Node<T>* newNode = new Node<T>(value);

        if (index == 0) {
            prepend(value);
            return;
        }

        Node<T>* prev = nodeAtIndex(index - 1);
        Node<T>* next = (prev != nullptr) ? prev->next : nullptr;

        if (prev != nullptr)
            prev->next = newNode;

        if (newNode != nullptr)
            newNode->next = next;
    }

    // Remove the node at a specific index in the linked list.
    void remove(int index) {
        if (index < 0)
            return;

        if (index == 0) {
            Node<T>* temp = head;
            head = head->next;
            delete temp;

            if (head == nullptr)
                tail = nullptr;

            return;
        }

        Node<T>* prev = nodeAtIndex(index - 1);
        if (prev == nullptr || prev->next == nullptr)
            return;

        Node<T>* toRemove = prev->next;
        prev->next = toRemove->next;
        delete toRemove;

        if (prev->next == nullptr)
            tail = prev;
    }

    // Print the linked list.
    void printList() const {
        Node<T>* current = head;

        while (current != nullptr) {
            std::cout << current->value << " -> ";
            current = current->next;
        }

        std::cout << "nullptr" << std::endl;
    }

    // Reverse the linked list.
    void reverse() {
        Node<T>* prev = nullptr;
        Node<T>* current = head;

        while (current != nullptr) {
            Node<T>* next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }

        tail = head;
        head = prev;
    }

    // Return the number of nodes in the linked list.
    int count() const {
        int count = 0;
        Node<T>* current = head;

        while (current != nullptr) {
            count++;
            current = current->next;
        }

        return count;
    }

    // Remove all nodes from the linked list.
    void removeAll() {
        Node<T>* current = head;

        while (current != nullptr) {
            Node<T>* next = current->next;
            delete current;
            current = next;
        }

        head = nullptr;
        tail = nullptr;
    }

    // Return a vector representation of the linked list.
    std::vector<T> toArray() const {
        std::vector<T> array;
        Node<T>* current = head;

        while (current != nullptr) {
            array.push_back(current->value);
            current = current->next;
        }

        return array;
    }

    // Remove all occurrences of a value from the linked list.
    void removeAllOccurrences(T value) {
        Node<T>* current = head;
        Node<T>* prev = nullptr;

        while (current != nullptr) {
            if (current->value == value) {
                if (prev != nullptr) {
                    prev->next = current->next;
                    delete current;
                    current = prev->next;
                } else {
                    Node<T>* temp = head;
                    head = head->next;
                    delete temp;
                    current = head;
                }
            } else {
                prev = current;
                current = current->next;
            }
        }

        if (head == nullptr)
            tail = nullptr;
    }

    // Perform an operation on each element of the linked list.
    void forEach(void (*operation)(T)) {
        Node<T>* current = head;

        while (current != nullptr) {
            operation(current->value);
            current = current->next;
        }
    }

    // Destructor to deallocate memory
    ~LinkedList() {
        Node<T>* current = head;
        Node<T>* next = nullptr;

        while (current != nullptr) {
            next = current->next;
            delete current;
            current = next;
        }
    }
};

int main() {
    LinkedList<int> list;
    list.append(1);
    list.append(2);
    list.append(3);

    std::cout << "Linked List: ";
    list.printList();

    return 0;
}
